// @shopify/shopify-app-session-storage-sqlite/build/ts/sqlite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteSessionStorage = void 0;
var tslib_1 = require("tslib");
var shopify_api_1 = require("@shopify/shopify-api");
var sqlite_connection_1 = require("./sqlite-connection");
var migrations_1 = require("./migrations");
var sqlite_migrator_1 = require("./sqlite-migrator");
var defaultSQLiteSessionStorageOptions = {
    sessionTableName: 'shopify_sessions',
    migratorOptions: {
        migrationDBIdentifier: 'shopify_sessions_migrations',
        migrationNameColumnName: 'migration_name',
    },
};
var SQLiteSessionStorage = /** @class */ (function () {
    function SQLiteSessionStorage(database, opts) {
        if (opts === void 0) { opts = {}; }
        this.options = tslib_1.__assign(tslib_1.__assign({}, defaultSQLiteSessionStorageOptions), opts);
        this.db = new sqlite_connection_1.SqliteConnection(database, this.options.sessionTableName);
        this.internalInit = this.init();
        this.migrator = new sqlite_migrator_1.SqliteSessionStorageMigrator(this.db, this.options.migratorOptions, migrations_1.migrationList);
        this.ready = this.migrator.applyMigrations(this.internalInit);
    }
    SQLiteSessionStorage.prototype.storeSession = function (session) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var entries, query;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        entries = session
                            .toPropertyArray()
                            .map(function (_a) {
                            var _b = tslib_1.__read(_a, 2), key = _b[0], value = _b[1];
                            return key === 'expires'
                                ? [key, Math.floor(value / 1000)]
                                : [key, value];
                        });
                        query = "\n      INSERT OR REPLACE INTO ".concat(this.options.sessionTableName, "\n      (").concat(entries.map(function (_a) {
                            var _b = tslib_1.__read(_a, 1), key = _b[0];
                            return key;
                        }).join(', '), ")\n      VALUES (").concat(entries
                            .map(function () { return "".concat(_this.db.getArgumentPlaceholder()); })
                            .join(', '), ");\n    ");
                        return [4 /*yield*/, this.db.query(query, entries.map(function (_a) {
                                var _b = tslib_1.__read(_a, 2), _key = _b[0], value = _b[1];
                                return value;
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.loadSession = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, rows, rawResult;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      SELECT * FROM ".concat(this.options.sessionTableName, "\n      WHERE id = ").concat(this.db.getArgumentPlaceholder(), ";\n    ");
                        return [4 /*yield*/, this.db.query(query, [id])];
                    case 2:
                        rows = _a.sent();
                        if (!Array.isArray(rows) || (rows === null || rows === void 0 ? void 0 : rows.length) !== 1)
                            return [2 /*return*/, undefined];
                        rawResult = rows[0];
                        return [2 /*return*/, this.databaseRowToSession(rawResult)];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.deleteSession = function (id) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      DELETE FROM ".concat(this.options.sessionTableName, "\n      WHERE id = ").concat(this.db.getArgumentPlaceholder(), ";\n    ");
                        return [4 /*yield*/, this.db.query(query, [id])];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.deleteSessions = function (ids) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      DELETE FROM ".concat(this.options.sessionTableName, "\n      WHERE id IN (").concat(ids
                            .map(function () { return "".concat(_this.db.getArgumentPlaceholder()); })
                            .join(','), ");\n    ");
                        return [4 /*yield*/, this.db.query(query, ids)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, true];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.findSessionsByShop = function (shop) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, rows, results;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ready];
                    case 1:
                        _a.sent();
                        query = "\n      SELECT * FROM ".concat(this.options.sessionTableName, "\n      WHERE shop = ").concat(this.db.getArgumentPlaceholder(), ";\n    ");
                        return [4 /*yield*/, this.db.query(query, [shop])];
                    case 2:
                        rows = _a.sent();
                        if (!Array.isArray(rows) || (rows === null || rows === void 0 ? void 0 : rows.length) === 0)
                            return [2 /*return*/, []];
                        results = rows.map(function (row) {
                            return _this.databaseRowToSession(row);
                        });
                        return [2 /*return*/, results];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.init = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var hasSessionTable, query;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.db.hasTable(this.options.sessionTableName)];
                    case 1:
                        hasSessionTable = _a.sent();
                        if (!!hasSessionTable) return [3 /*break*/, 3];
                        query = "\n        CREATE TABLE ".concat(this.options.sessionTableName, " (\n          id varchar(255) NOT NULL PRIMARY KEY,\n          shop varchar(255) NOT NULL,\n          state varchar(255) NOT NULL,\n          isOnline integer NOT NULL,\n          expires integer,\n          scope varchar(1024),\n          accessToken varchar(255),\n          onlineAccessInfo varchar(255)\n        );\n      ");
                        return [4 /*yield*/, this.db.query(query)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SQLiteSessionStorage.prototype.databaseRowToSession = function (row) {
        // convert seconds to milliseconds prior to creating Session object
        if (row.expires)
            row.expires *= 1000;
        return shopify_api_1.Session.fromPropertyArray(Object.entries(row));
    };
    return SQLiteSessionStorage;
}());
exports.SQLiteSessionStorage = SQLiteSessionStorage;
