// @shopify/app/dist/cli/services/generate/fetch-template-specifications.js
import { RemoteTemplateSpecificationsQuery, } from '../../api/graphql/template_specifications.js';
import { BaseFunctionConfigurationSchema } from '../../models/extensions/schemas.js';
import { blocks, templates } from '../../constants.js';
import { partnersRequest } from '@shopify/cli-kit/node/api/partners';
export async function fetchTemplateSpecifications(token) {
    const params = { version: templates.specification.remoteVersion };
    const result = await partnersRequest(RemoteTemplateSpecificationsQuery, token, params);
    return result.templateSpecifications.map(mapRemoteTemplateSpecification);
}
export function mapRemoteTemplateSpecification(remoteTemplateSpecification) {
    return {
        identifier: remoteTemplateSpecification.identifier,
        name: remoteTemplateSpecification.name,
        group: remoteTemplateSpecification.group,
        supportLinks: remoteTemplateSpecification.supportLinks,
        types: remoteTemplateSpecification.types.map((extension) => {
            return {
                identifier: remoteTemplateSpecification.identifier,
                externalIdentifier: remoteTemplateSpecification.identifier,
                externalName: remoteTemplateSpecification.identifier,
                gated: false,
                registrationLimit: blocks.functions.defaultRegistrationLimit,
                supportedFlavors: extension.supportedFlavors,
                group: remoteTemplateSpecification.group,
                category: () => 'function',
                configSchema: BaseFunctionConfigurationSchema,
                templateURL: extension.url,
                helpURL: remoteTemplateSpecification.supportLinks[0],
                templatePath: (flavor) => {
                    const supportedFlavor = extension.supportedFlavors.find((supportedFlavor) => supportedFlavor.value === flavor);
                    if (!supportedFlavor)
                        return undefined;
                    return supportedFlavor.path;
                },
            };
        }),
    };
}
