// @shopify/shopify-app-session-storage/build/cjs/abstract-migration-engine.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var types = require('./types.js');

class AbstractMigrationEngine {
  constructor(db, opts = {}, migrations) {
    this.options = void 0;
    this.connection = void 0;
    this.ready = void 0;
    this.migrations = void 0;
    this.options = {
      ...types.defaultSessionStorageMigratorOptions,
      ...opts
    };
    this.connection = db;
    this.migrations = migrations;
    this.ready = this.initMigrationPersistence();
  }
  async applyMigrations(databaseReady) {
    await databaseReady;
    await this.ready;
    for (const {
      migrationName,
      migrationFunction
    } of this.getMigrationList()) {
      const migrationApplied = await this.hasMigrationBeenApplied(migrationName);
      if (!migrationApplied) {
        await migrationFunction(this.connection);
        await this.saveAppliedMigration(migrationName);
      }
    }
    return Promise.resolve();
  }
  getMigrationList() {
    return this.migrations;
  }
}

exports.AbstractMigrationEngine = AbstractMigrationEngine;
